lib/src/http/base_request.dart

```diff
@@ -0,0 +1,70 @@
+library w_transport.src.http.base_request;
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:fluri/fluri.dart';
+import 'package:http_parser/http_parser.dart';
+
+import 'package:w_transport/src/http/request_dispatchers.dart';
+import 'package:w_transport/src/http/request_progress.dart';
+
+/// A common API that applies to all request types. The piece that is missing is
+/// that which is specific to the request body. Setting the request body differs
+/// based on the type of request being sent (plain-text, JSON, form, multipart,
+/// or streamed). As such, that piece of the API is delegated to the specific
+/// request classes.
+abstract class BaseRequest implements FluriMixin, RequestDispatchers {
+  /// Gets and sets the content-length of the request, in bytes. If the size of
+  /// the request is not known in advance, set this to null.
+  int contentLength;
+
+  /// Content-type of this request.
+  ///
+  /// By default, the mime-type is "text/plain" and the charset is "UTF8".
+  /// When the request body or the encoding is set or updated, the content-type
+  /// will be updated accordingly.
+  MediaType get contentType;
+
+  /// Future that resolves when the request has completed (successful or
+  /// otherwise).
+  Future<Null> get done;
```

Comment:
Yes, because the Future will always complete with null. It's just being explicit about not returning anything so the consumer doesn't mistake it with untyped or forgotten